### PR TITLE
[Refactor] Add IMaterializedViewRewriter and IMaterializedViewRewriteRule to make it more extensible

### PR DESCRIPTION
**Pull request by LiShuMing**
## Why I'm doing:
Split PR (https://github.com/StarRocks/starrocks/pull/42809).

## What I'm doing:
1. Add IMaterializedViewRewriter and IMaterializedViewRewriteRule to make it more extensible
2. Move `HLLRewriteEquivalent` and `PercentileRewriteEquivalent` from `AggregateFunctionRewriter` to make it more extensible.
3. Refactor other codes.

Fixes #issue

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [ ] Enhancement
- [x] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.3
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by LiShuMing** [comment]
@mergify rebase

**Review by Seaven — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java [hunk 1]
```diff
@@ -75,27 +71,13 @@ public boolean canRewriteAggFunction(ScalarOperator op) {
         if (aggFuncName.equals(FunctionSet.AVG)) {
             return true;
         }
-        // HLL
-        if (aggFuncName.equals(FunctionSet.APPROX_COUNT_DISTINCT) || aggFuncName.equals(FunctionSet.NDV) ||
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/AggregateFunctionRewriter.java:79]: why not refactor avg into IRewriteEquivalent?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+/**
+ * Base interface to implement materialized view rewrite rule.
+ */
+public interface IMaterializedViewRewriter {
+    /**
+     * Rewrite the query with the given materialized view context.
+     * @param mvContext: materialized view context of query and associated mv.
+     * @return: the rewritten query opt expression if rewrite success, otherwise null.
+     */
+    OptExpression doRewrite(MvRewriteContext mvContext);
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java:30]:
```suggestion
    OptExpression rewrite(MvRewriteContext mvContext);
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+
+    /**
+     * After plan is rewritten by MV, still do some actions for new MV's plan.
+     * 1. column prune
+     * 2. partition prune
+     * 3. bucket prune
+     */
+    OptExpression doPostAfterRewrite(OptimizerContext optimizerContext,
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java:38]:
```suggestion
    OptExpression doAfterRewrite(OptimizerContext optimizerContext,
    or
    OptExpression postRewrite(OptimizerContext optimizerContext,
```

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/IMaterializedViewRewriteRule.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+ */
+public interface IMaterializedViewRewriteRule {
+    /**
+     * Prune the materialized view candidates.
+     * @param queryExpression: query opt expression.
+     * @param context: optimizer context.
+     * @param mvCandidateContexts: materialized view candidates prepared in the mv preprocessor.
+     * @return: the pruned materialized view candidates.
+     */
+    List<MaterializationContext> doPrune(OptExpression queryExpression,
```
Comment (ABingHuang) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/rule/IMaterializedViewRewriteRule.java:36]:
```suggestion
    List<MaterializationContext> pruneCandidates(OptExpression queryExpression,
```

**Review by ABingHuang — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java (hunk 3 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/materialization/IMaterializedViewRewriter.java:38]: will do this in next pr.

**Pull request merged**